0016: update KBI to include `datalad diff` (#81)
psychoinformatics-de/knowledge-base: kbi/0016/index.rst

to the dataset by a specific commit, while all other files that are available locally
should remain untouched.

This Knowledge Base Item outlines several methods for dropping local files that were added
in a specific commit. These methods differ in the way they identify which files to drop
(via ``datalad status``, ``git diff``, or ``datalad diff``), but the actual dropping of
content is handled by ``datalad drop`` in all cases.

Content
-------

- `Preparation`_
- `Using datalad status`_
- `Using git diff`_
- `Dropping the files`_
- `Using datalad diff in a one-liner`_
- `Drop limitation`_

> *Note: if you are not interested in details and just looking for the quickest and leanest
> way to get the job done, skip over to the section: `Using datalad diff in a one-liner`_*

Reviewer: ```suggestion
.. note::
   If you are not interested in details and just looking for the quickest and leanest
   way to get the job done, skip over to the section: `Using datalad diff in a one-liner`_*.
```
The link didn't render for me - I don't know how to make inline emphasis and hyperlink markup play nicely together, but maybe a note directive would look nicely?

to the dataset by a specific commit, while all other files that are available locally
should remain untouched.

This Knowledge Base Item outlines several methods for dropping local files that were added
in a specific commit. These methods differ in the way they identify which files to drop
(via ``datalad status``, ``git diff``, or ``datalad diff``), but the actual dropping of
content is handled by ``datalad drop`` in all cases.

Content
-------

- `Preparation`_
- `Using datalad status`_
- `Using git diff`_
- `Dropping the files`_
- `Using datalad diff in a one-liner`_
- `Drop limitation`_

> .. note::
>    If you are not interested in details and just looking for the quickest and leanest
>    way to get the job done, skip over to the section: `Using datalad diff in a one-liner`_*.

Author: ```suggestion
.. note::

   If you are not interested in details and just looking for the quickest and leanest
   way to get the job done, skip over to the section: `Using datalad diff in a one-liner`_.
```

   {"action": "drop", "annexkey": "MD5E-s10--6fe97938d91d6a56a50c14caa5c81e12.txt", "path": "/Users/jsheunis/Documents/psyinf/Data/drop-files-test/file2.txt", "refds": "/Users/jsheunis/Documents/psyinf/Data/drop-files-test", "status": "ok", "type": "file"}
   {"action": "drop", "annexkey": "MD5E-s10--6fe97938d91d6a56a50c14caa5c81e12.txt", "path": "/Users/jsheunis/Documents/psyinf/Data/drop-files-test/file3.txt", "refds": "/Users/jsheunis/Documents/psyinf/Data/drop-files-test", "status": "ok", "type": "file"}

.. _Using datalad diff in a one-liner:

Using ``datalad diff`` in a one-liner
-------------------------------------

`datalad diff`_ provides similar information as ``git diff``, although with additonial
functionality related to (nested) DataLad datasets.

If you enjoy running one-liners and preventing unnecessary write operations to disk,
this option is for you. Below is a single line of code that uses ``datalad diff``, 
``datalad drop``, and standard UNIX tools to identify and drop files related to a
specific commit:

.. code-block:: console

   datalad drop $(datalad -f '{state}:{path}' diff -f HEAD~1 -t HEAD | grep '^added:' | cut -d ':' -f 2-)

To explain:

> - ``-f 42f197501c3293bc1c0c22e36b1618eec706090e -t 73489f56ecd5eb4dee14c957349f09c0d8b1684d``
>   uses ``datalad diff``'s ``--from`` and ``--to`` options to specify the two states that
>   will be compared (here using the commit shasums). Alternatively, symbolic names could
>   also be used, for example ``-f HEAD~1 -t HEAD`` to refer to the last commit.
> - ``-f '{state}:{path}'`` uses DataLad's custom formatting option to format results of
>   the ``datalad diff`` command. It produces output like
>   ``added::/Users/jsheunis/Documents/psyinf/Data/drop-files-test/file2.txt``.
> - ``grep`` and ``cut`` are standard UNIX tools to find lines that start with ``added:``,
>   and to only report on the path that is contained in these lines.

Reviewer: This may be pretty hard to read in GitHubs change diff, but I'd like to suggest changes that:

- Swap irst two bullet points to explain parameters in the order they are used
- Use `-f HEAD~1` in explanation and present shasum as an alternative (example directly above uses `HEAD~1`).
- Reword things a little

```suggestion
- ``-f '{state}:{path}'`` provides an output format template which will be used to
  format results of the ``datalad diff`` command. It produces output like
  ``added::/Users/jsheunis/Documents/psyinf/Data/drop-files-test/file2.txt``.
- ``-f HEAD~1 -t HEAD`` uses ``datalad diff``'s ``--from`` and ``--to`` options
  to specify the two states that will be compared (here using symbolic names referring
  to previous and last commit). Full or partial commit shasums can also be used like in
  previous examples (``-f 42f197501c3293bc1c0c22e36b1618eec706090e -t
  73489f56ecd5eb4dee14c957349f09c0d8b1684d``)
- ``grep`` and ``cut`` are standard UNIX tools; here they are used to find lines starting
  with ``added:``, and to extract only the path that is contained in these lines.
```
